src/upsample.jl

```diff
@@ -64,235 +64,175 @@ function ChainRulesCore.rrule(::typeof(upsample_nearest), x::AbstractArray, s::T
-    return y
-end
-
-function get_inds_and_ws(x::T, n::Int, m::Int, dim::Int) where T <: AbstractArray
-    # Creates interpolation grid for resampling. 
-    # Creates the same grid as used in Image.jl `imresize`.
-    step = n // m
-    offset = (n + 1)//2 - step//2 - step * (m//2 - 1)
-    xq = clamp.(range(offset, step=step, length=m), 1, n)
-
-    # Creates interpolation lower and upper indices, and broadcastable weights
-    ilow = floor.(Int, xq)
-    ihigh = ceil.(Int, xq)
-    sizew = ntuple(i-> i == dim ? length(xq) : 1, ndims(x))
-    wdiff = convert(T, reshape(xq .- ilow, sizew)) # wdiff possibly lives on gpu
-    return ilow, ihigh, wdiff
+# utility function
+@inline function compute_source_index_and_lambda(
+    ratio, # 0 < ratio < 1
+    output_index,
+    input_size,
+    output_size
+)
+    real_input_index = ratio*output_index
+    input_index0 = floor(Int, real_input_index) # typecast to int was here in C++
+    offset = (input_index0 < input_size - 1) ? 1 : 0
+    input_index1 = input_index0 + offset
+    lambda1 = real_input_index - input_index0
+    lambda0 = 1 - lambda1
+    return input_index0, input_index1, lambda0, lambda1
 end
 
 """
-    adjoint_of_idx(idx::Vector{<:Integer})
-
-# Arguments
-- `idx`: a vector of indices from which you want the adjoint.
+    upsample_bilinear(x::AbstractArray{T,4}, scale::NTuple{2,Real}=(1,1); size::Union{Nothing,NTuple{2,Integer}}=nothing)
 
-# Outputs
--`idx_adjoint`: index that inverses the operation `x[idx]`.
+Upsamples the first 2 dimensions of the array `x` by the upsample factors stored in `scale`,
+using bilinear interpolation.
 
-# Explanation
-Determines the adjoint of the vector of indices `idx`, based on the following assumptions:
-* `idx[1] == 1`
-* `all(d in [0,1] for d in diff(idx))`
-The adjoint of `idx` can be seen as an inverse operation such that:
+The size of the output is equal to
+`(scale[1]*S1, scale[2]*S2, S3, S4)`, where `S1, S2, S3, S4 = size(x)`.
 
+Examples:
 ```julia
-x = [1, 2, 3, 4, 5]
-idx = [1, 2, 2, 3, 4, 4, 5]
-idx_adjoint = adjoint_of_idx(idx)
-@assert x[idx][idx_adjoint] == x
+upsample_bilinear(x, (2, pi)) # real scaling factors are allowed
+upsample_bilinear(x; size=(64,64)) # note the semicolon, size is a keyword argument
 ```
-The above holds as long as `idx` contains every index in `x`.
+Currently only 2d upsampling is supported.
 """
-function adjoint_of_idx(idx::Vector{Int})
-    d = trues(length(idx))
-    d[2:end] .= diff(idx)
-    idx_adjoint = findall(d)
-    return idx_adjoint
+function upsample_bilinear(x::AbstractArray{T,4}, scale::NTuple{2,Real}=(1,1); size::Union{Nothing,NTuple{2,Integer}}=nothing) where T
+    w,h,c,n = Base.size(x)
+    if scale != (1,1) && size !== nothing
+        error("Please provide either scale or size, not both. Got scale=$scale and size=$size.")
+    end
+    if size === nothing
+        out_w = floor(Int, scale[1]*w)
+        out_h = floor(Int, scale[2]*h)
+    else
+        out_w, out_h = size
+    end
```

Comment:
I don't see why the test fails. But one comment here is that using `(1,1)` as a sentinel value might be confusing, I mean no human will deliberately write that but some program might. Perhaps they should both be `nothing`?

```suggestion
function upsample_bilinear(x::AbstractArray{T,4}, scale::Union{Nothing,NTuple{2,Real}}=nothing; size::Union{Nothing,NTuple{2,Integer}}=nothing) where {T}
    w,h,c,n = Base.size(x)
    if size === nothing && scale !== nothing
        out_w = floor(Int, scale[1]*w)
        out_h = floor(Int, scale[2]*h)
    elseif size !== nothing && scale === nothing
        out_w, out_h = size
    else
        throw(ArgumentError("Please provide either scale or size, not both. Got scale=$scale and size=$size."))
    end
```

Reply:
I think if some program sets `scale=(1,1)` it wouldn't change anything either way, or am I missing sth? Having `(1,1)` as default is just shorter and omits the union, so I'd leave it as is.

Reply:
Ah you are right, it should run fine with `(1,1)`. The only thing it won't do is throw an error with both (1,1) and a keyword, it will silently choose the keyword, but that's a pretty far-out edge case, no problem.